k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +106,324 @@ class SingleNsNamer(
+    residual: Path
+  )(lookup: Option[Var[Set[Address]]]): NameTree[Name] = lookup match {
+    case Some(addresses) =>
+      val addrs = addresses.map { Addr.Bound(_) }
+      NameTree.Leaf(Name.Bound(addrs, idPrefix ++ id, residual))
+    case None => NameTree.Neg
   }
 
-  override val getAllNames: Activity[Set[Path]] = {
-    // explicit type annotations are required for scala to pick the right
-    // versions of flatMap and map
-    val namespaces: ActSet[String] = Activity(endpointNs.namespaces.map(Activity.Ok(_)))
-    namespaces.flatMap { namespace: String =>
-      val services: ActSet[SvcCache] = endpointNs.get(namespace, None).services.map(_.values.toSet)
-      services.flatMap { service: SvcCache =>
-        val ports: Var[Set[String]] = service.ports.map(_.keys.toSet)
-        val states = ports.map { ports =>
-          val paths = ports.map { port =>
-            idPrefix ++ Path.Utf8(namespace, port, service.name)
+  private[k8s] def lookupServices(
+    nsName: String,
+    portName: String,
+    serviceName: String,
+    id: Path,
+    residual: Path,
+    labelSelector: Option[String] = None
+  ): Activity[NameTree[Name]] = {
+    val endpointsAct = serviceEndpoints(nsName, serviceName, labelSelector)
+    val unstable = Try(portName.toInt).toOption match {
+      case Some(portNumber) =>
+        // if `portName` was successfully parsed as an `int`, then
+        // we are dealing with a numbered port. we will thus also
+        // need the port mappings from the `Service` API response,
+        // so join its activity with the endpoints activity.
+        endpointsAct.join(numberedPortRemappings(nsName, serviceName, labelSelector))
+          .map { case (endpoints, ports) =>
+            endpoints.lookupNumberedPort(ports, portNumber)
           }
-          Activity.Ok(paths)
-        }
-        Activity(states)
-      }
+      case None =>
+        // otherwise, we are dealing with a named port, so we can
+        // just look up the service from the endpoints activity
+        endpointsAct.map { endpoints => endpoints.lookupNamedPort(portName) }
     }
+    stabilize(unstable).map { mkNameTree(id, residual) }
   }
 }
 
-private object EndpointsNamer {
-  case class Endpoint(ip: InetAddress, nodeName: Option[String])
+object EndpointsNamer {
+  val DefaultBackoff: Stream[Duration] =
+    Backoff.exponentialJittered(10.milliseconds, 10.seconds)
 
-  case class Svc(endpoints: Set[Endpoint], ports: Map[String, Int])
+  protected type PortMap = Map[String, Int]
+  protected type NumberedPortMap = Map[Int, String]
 
-  private[this] def getEndpoints(subsets: Option[Seq[v1.EndpointSubset]]): Set[Endpoint] = {
-    val endpoints = mutable.Set.empty[Endpoint]
-    for {
-      subset <- subsets.getOrElse(Seq.empty)
-      addresses <- subset.addresses
-      addrs <- addresses
-    } {
-      endpoints += Endpoint(InetAddress.getByName(addrs.ip), addrs.nodeName)
-    }
-    endpoints.toSet
+  private trait Loggable {
```

Comment:
Lots of changes to the logging logic in this branch.  Do you have any ideas about the best way to test these changes?